### PR TITLE
Resolve dependency using entry config

### PR DESCRIPTION
**Pull request by yufeih**
Fixes https://dev.azure.com/ceapex/Engineering/_workitems/edit/113864

The root cause is we use `Docset.DependencyDocsets` to lookup dependencies, which is empty when `Docset` itself is a dependency docset. This PR moves `Docset.DependencyDocsets` to dependency resolver and uses the same `_dependencies` dictionary to resolve everything.

Due to this change, we always use entry docset config for everything including dependencies, so some restore tests are marked as skip because they specify a config in dependency docset which is currently a not supported scenario in the docs world. We'll revisit these scenarios when doing `dotnet-docs` which uses _content from CRR_.

Dependency resolve and fallback resolve is consolidated into the same place at `DependencyResolver.TryResolveFile` to clarify the logic.

Simplified unnecessary Docset constructor parameters.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5003)

### CODE REVIEW

**Comment by OsmondJiang** [comment]
Can we just remove the error code "Link Is Dependency" and use  LinkOutOfScope only? Then we can easily fix this problem at:

image

**Comment by yufeih** [comment]
For resources, we currently build resources from main repo but is outside build scope, and so we don't build resources from dependency repo (outside build scope), so need a flag to tell if a file is from a dependency repo or not.

**Comment by OsmondJiang** [comment]
> For resources, we currently build resources from main repo but is outside build scope, and so we don't build resources from dependency repo (outside build scope), so need a flag to tell if a file is from a dependency repo or not.

Then we can just use the _buildScope to tell us if it's dependency docset or not. we don't need change so much ....

**Comment by OsmondJiang** [comment]
As we talked offline, use the same _dependencyDocsets to resolve everything is not correct, the _dependencyDocsets need follows the Docset, no matter it's entry docset or dependency docset.

File: docs/specs/restore.yml [hunk 1]
```diff
@@ -184,7 +185,7 @@ repos:
 outputs:
   docs/a.json:
 ---
-# restore with dependency lock
+# [skip] restore with dependency lock
```
Comment (OsmondJiang) [inline docs/specs/restore.yml]: still skipped?

File: src/docfx/build/context/BuildScope.cs [hunk 2]
```diff
@@ -27,19 +25,23 @@ internal class BuildScope
 
         public RedirectionMap Redirections { get; }
 
+        public Docset Docset { get; }
+
+        public Docset FallbackDocset { get; }
```
Comment (OsmondJiang) [inline src/docfx/build/context/BuildScope.cs]: I try to hide FallbackDocset from outside of build scope, can we still keep it hidden?

File: src/docfx/build/dependency/DependencyResolver.cs [hunk 3]
```diff
@@ -280,38 +283,83 @@ private string ResolveToDocsetRelativePath(string path, Document referencingFile
             return docsetRelativePath;
         }
 
-        private Document TryResolveResourceFromHistory(
-            Document referencingFile, GitCommitProvider gitCommitProvider, string pathToDocset, TemplateEngine templateEngine)
+        private Document TryResolveFile(Document declaringFile, string pathToDocset)
```
Comment (live1206) [inline src/docfx/build/dependency/DependencyResolver.cs]:
```suggestion
        private Document TryResolveFile(Document referencingFile, string pathToDocset)
```

File: src/docfx/build/dependency/DependencyResolver.cs [hunk 4]
```diff
@@ -18,15 +18,17 @@ internal class DependencyResolver
         private readonly DependencyMapBuilder _dependencyMapBuilder;
         private readonly GitCommitProvider _gitCommitProvider;
         private readonly IReadOnlyDictionary<string, string> _resolveAlias;
+        private readonly IReadOnlyDictionary<string, Docset> _dependencies;
```
Comment (OsmondJiang) [inline src/docfx/build/dependency/DependencyResolver.cs]: >_dependencies [](start = 61, length = 13)

Is it better to put it in BuildScope?

File: src/docfx/build/dependency/DependencyResolver.cs [hunk 5]
```diff
@@ -280,38 +283,83 @@ private string ResolveToDocsetRelativePath(string path, Document referencingFile
             return docsetRelativePath;
         }
 
-        private Document TryResolveResourceFromHistory(
-            Document referencingFile, GitCommitProvider gitCommitProvider, string pathToDocset, TemplateEngine templateEngine)
+        private Document TryResolveFile(Document referencingFile, string pathToDocset)
```
Comment (OsmondJiang) [inline src/docfx/build/dependency/DependencyResolver.cs]: >Document [](start = 16, length = 8)

Also for this one

File: src/docfx/build/metadata/ContributionProvider.cs [hunk 6]
```diff
@@ -196,7 +197,7 @@ string GetContentGitUrl(string branchUrlTemplate)
                     (branchUrlTemplate, _) = GetContentGitUrlTemplate(contributionRemote, pathToRepo);
 
                     (editRemote, editBranch) = (contributionRemote, hasRefSpec ? contributionBranch : editBranch);
-                    if (context.BuildScope.GetFallbackDocset(document.Docset) != null)
```
Comment (OsmondJiang) [inline src/docfx/build/metadata/ContributionProvider.cs]: >context [](start = 24, length = 7)

FilePath.From

**Pull request merged**